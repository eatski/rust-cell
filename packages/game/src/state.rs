use std::{
    collections::{BTreeMap, BTreeSet},
    ops::Add,
};

#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy, PartialOrd, Ord)]
pub struct Address {
    pub x: isize,
    pub y: isize,
}

#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy, PartialOrd, Ord)]
pub struct RelativePath {
    pub x: isize,
    pub y: isize,
}

impl Add<&RelativePath> for &RelativePath {
    type Output = RelativePath;
    fn add(self, rhs: &RelativePath) -> Self::Output {
        RelativePath {
            x: self.x + rhs.x,
            y: self.y + rhs.y,
        }
    }
}

pub const NEXT_PATHES: [RelativePath; 4] = [
    RelativePath { x: 0, y: -1 },
    RelativePath { x: 0, y: 1 },
    RelativePath { x: -1, y: 0 },
    RelativePath { x: 1, y: 0 },
];

pub const UNIT_CORE_PATH: RelativePath = RelativePath { x: 0, y: 0 };

impl Add<&RelativePath> for &Address {
    type Output = Address;
    fn add(self, rhs: &RelativePath) -> Self::Output {
        Address {
            x: self.x + rhs.x,
            y: self.y + rhs.y,
        }
    }
}

#[derive(Debug, Clone)]
pub struct Unit {
    pub pathes: BTreeSet<RelativePath>,
    pub blueprint: BTreeSet<RelativePath>,
}

impl Default for Unit {
    fn default() -> Self {
        Self {
            pathes: [UNIT_CORE_PATH].into(),
            blueprint: Default::default(),
        }
    }
}

impl Unit {
    pub fn new(blueprint: BTreeSet<RelativePath>) -> Self {
        Self {
            pathes: [UNIT_CORE_PATH].into(),
            blueprint,
        }
    }
}

pub type UnitId = usize;

#[derive(Debug)]
pub struct FinalizedGameState {
    pub cells: Vec<(Address, Unit)>,
}

#[derive(Debug, Default)]
pub struct GameState {
    pub units: BTreeMap<UnitId, (Address, Unit)>,
    pub cells: BTreeMap<Address, (UnitId, RelativePath)>,
}

impl GameState {
    pub fn dry_run_move_unit<'a>(
        &'a mut self,
        unit_id: &'a UnitId,
        direction: &'a RelativePath,
    ) -> Result<impl FnOnce() + 'a, BTreeSet<(UnitId, RelativePath)>> {
        let (address, unit) = self.units.get(unit_id).unwrap().clone();
        let current_pathes = unit.pathes.clone();
        let current_address: BTreeSet<_> = current_pathes
            .into_iter()
            .map(|path| (&address + &path, path))
            .collect();
        let next_addresses: BTreeSet<_> = current_address
            .iter()
            .map(move |(address, path)| (address + &direction, path.clone()))
            .collect();

        let collision_addresses: BTreeSet<_> = next_addresses
            .iter()
            .filter_map(|(address, _)| {
                let next_unit_id = self.cells.get(&address);
                next_unit_id.and_then(|(next_unit_id, path)| {
                    if next_unit_id != unit_id {
                        Some((next_unit_id.clone(), path.clone()))
                    } else {
                        None
                    }
                })
            })
            .collect();
        if collision_addresses.is_empty() {
            Ok(move || {
                let (address, _) = self.units.get_mut(&unit_id).unwrap();
                for (address, _) in current_address {
                    self.cells.remove(&address);
                }
                for (address, path) in next_addresses {
                    self.cells.insert(address, (*unit_id, path.clone()));
                }
                *address = &address.clone() + &direction;
            })
        } else {
            Err(collision_addresses)
        }
    }

    /**
     * unitを新しくスポーンさせる
     */
    pub fn dry_run_spawn_unit<'a>(
        &'a mut self,
        address: &'a Address,
    ) -> Option<impl FnOnce(Unit) + 'a> {
        if self.cells.contains_key(address) {
            return None;
        }
        let address = address.clone();
        Some(move |unit| {
            let unit_id = self.units.keys().max().unwrap_or(&0) + 1;
            self.cells.insert(address, (unit_id, UNIT_CORE_PATH));
            self.units.insert(unit_id, (address, unit));
        })
    }

    /**
     * unitのpathを追加する
     */
    pub fn dry_run_add_path<'a>(
        &'a mut self,
        unit_id: &UnitId,
        path: &RelativePath,
    ) -> Option<impl FnOnce() + 'a> {
        let new_address = {
            let (address, _) = self.units.get(unit_id).unwrap();
            &address.clone() + path
        };
        let new_address_cloned = new_address.clone();
        if self.cells.contains_key(&new_address) {
            return None;
        }
        let path = path.clone();
        let unit_id = unit_id.clone();
        Some(move || {
            let (_, unit) = self.units.get_mut(&unit_id).unwrap();
            unit.pathes.insert(path.clone());
            self.cells.insert(new_address_cloned, (unit_id, path));
        })
    }

    /**
     * unitを削除する
     */
    pub fn remove_unit(&mut self, unit_id: &UnitId) {
        let (address, unit) = self.units.remove(unit_id).unwrap();
        for path in unit.pathes {
            self.cells.remove(&(&address + &path));
        }
    }

    pub fn finalize(&self) -> FinalizedGameState {
        let cells = self
            .units
            .iter()
            .map(|(_, (address, unit))| (address.clone(), unit.clone()))
            .collect();
        FinalizedGameState { cells }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn move_unit() {
        let mut state = GameState::default();
        state.dry_run_spawn_unit(&Address { x: 0, y: 0 }).unwrap()(Unit::default());
        let unit_id = state.units.first_key_value().unwrap().0.clone();
        state.dry_run_spawn_unit(&Address { x: 2, y: 0 }).unwrap()(Unit::default());
        state
            .dry_run_move_unit(&unit_id, &RelativePath { x: 1, y: 0 })
            .unwrap()();
        insta::assert_debug_snapshot!(state.finalize());
    }

    #[test]
    fn move_unit_2() {
        let mut state = GameState::default();
        state.dry_run_spawn_unit(&Address { x: 0, y: 0 }).unwrap()(Unit::default());
        let unit_id = state.units.first_key_value().unwrap().0.clone();
        state.dry_run_spawn_unit(&Address { x: 2, y: 0 }).unwrap()(Unit::default());
        state
            .dry_run_move_unit(&unit_id, &RelativePath { x: -1, y: 0 })
            .unwrap()();
        insta::assert_debug_snapshot!(state.finalize());
    }

    #[test]
    fn move_unit_3() {
        let mut state = GameState::default();
        state.dry_run_spawn_unit(&Address { x: 0, y: 0 }).unwrap()(Unit::default());
        let unit_id = state.units.first_key_value().unwrap().0.clone();
        state.dry_run_spawn_unit(&Address { x: 1, y: 0 }).unwrap()(Unit::default());
        let error = state
            .dry_run_move_unit(&unit_id, &RelativePath { x: 1, y: 0 })
            .map(|_| ())
            .unwrap_err();
        insta::assert_debug_snapshot!(error);
    }
}
