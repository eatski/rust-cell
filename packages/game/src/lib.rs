pub mod state;
use std::collections::BTreeSet;

use rand::{
    seq::{IteratorRandom, SliceRandom},
    Rng,
};
use state::*;

pub enum Input {
    Click { address: Address },
}

/**
 * 次の点を計算する
 * 点はランダムに上下左右に1単位動く
 * inputsがある場合は、入力に応じて点を追加する
 */
pub fn update(state: &mut GameState, inputs: &Vec<Input>, rng: &mut impl Rng) {
    let mut units_to_iter: Vec<_> = state.units.clone().into_iter().collect();
    units_to_iter.shuffle(rng);
    for (current_unit_id, _) in units_to_iter.iter() {
        let rnd = rng.gen_range(0..2048);
        match rnd {
            0..=24 => {
                let direction = NEXT_PATHES.iter().choose(rng).unwrap();
                let fail = {
                    match state.dry_run_move_unit(current_unit_id, direction) {
                        Ok(exec) => {
                            exec();
                            None
                        }
                        Err(collisions) => Some(collisions),
                    }
                };
                if let Some(collisions) = fail {
                    let (cores, others): (Vec<_>, Vec<_>) = collisions
                        .into_iter()
                        .partition(|(_, path)| path == &UNIT_CORE_PATH);
                    if others.is_empty() {
                        // coresのみ衝突している場合は、衝突したunitを消す
                        for (id, _) in cores {
                            state.remove_unit(&id);
                        }
                        state.dry_run_move_unit(current_unit_id, direction).unwrap()();
                    }
                }
            }
            25 => {
                let (_, unit) = state.units.get(current_unit_id).unwrap();
                let next_pathes: BTreeSet<_> = unit
                    .pathes
                    .iter()
                    .flat_map(|path| NEXT_PATHES.iter().map(move |next_path| path + next_path))
                    .collect();
                let mut next_pathes: Vec<_> = next_pathes.into_iter().collect();
                next_pathes.shuffle(rng);

                for path in next_pathes.iter() {
                    if let Some(exec) = state.dry_run_add_path(current_unit_id, path) {
                        exec();
                        break;
                    }
                }
            }
            _ => {}
        }
    }
    if let Some(Input::Click { address }) = inputs.last() {
        if let Some(_) = state.cells.get(address).cloned() {
        } else {
            state.spawn_unit(address);
        }
    }
}

#[cfg(test)]
mod update_test {
    use super::*;
    use rand::rngs::StdRng;
    use rand::SeedableRng;
    #[test]
    fn update_test() {
        let seed: [u8; 32] = [
            1, 8, 13, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24,
            25, 26, 27, 28, 29, 30, 31, 32,
        ];
        let mut rng: StdRng = SeedableRng::from_seed(seed);
        let mut state = Default::default();

        for _ in 0..30 {
            let address = Address {
                x: rng.gen_range(0..100),
                y: rng.gen_range(0..100),
            };
            update(&mut state, &vec![Input::Click { address }], &mut rng);
        }
        for _ in 0..50 {
            update(&mut state, &vec![], &mut rng);
        }
        insta::assert_debug_snapshot!(state.finalize());
        for _ in 0..50 {
            update(&mut state, &vec![], &mut rng);
        }
        insta::assert_debug_snapshot!(state.finalize());
        for _ in 0..10000 {
            update(&mut state, &vec![], &mut rng);
        }
        insta::assert_debug_snapshot!(state.finalize());
    }
}
