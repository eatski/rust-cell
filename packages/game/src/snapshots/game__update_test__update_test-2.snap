---
source: src/lib.rs
expression: state
---
GameState {
    cells: {
        Address {
            x: -1,
            y: 3,
        }: 2,
        Address {
            x: 0,
            y: -1,
        }: 1,
        Address {
            x: 4,
            y: 0,
        }: 3,
    },
    units: {
        1: Unit {
            order: None,
        },
        2: Unit {
            order: None,
        },
        3: Unit {
            order: None,
        },
    },
}
